### ISSUE TITLE
[HttpClient] Can we not trigger_error and throw exception or log warning instead?

### ISSUE THREAD

**Issue by simPod** [symfony/symfony#42100]
**Description**
StreamWrapper triggers error (lvl warning).

https://github.com/symfony/symfony/blob/b5d4b33d5e65d956194c170fda5f8f40bc6a921a/src/Symfony/Component/HttpClient/Response/StreamWrapper.php#L185

This causes an issue because it always falls into an error handler that is currently set (`set_error_handler()`).

So even though I catch the error using `catch(ErrorException){ }`, the error handler is triggered anyways.

**Example**
Sentry php sdk sets its own error handler. `trigger_error()` triggers it without me having any control over it. So I have to unset the error handler before making a request and then set it back again.

**Comment by derrabus** [comment]
See https://www.php.net/manual/en/streamwrapper.stream-read.php

Triggering a warning is the expected behavior of a stream wrapper implementation if the stream cannot be read.

**Comment by simPod** [comment]
Hmm, I see. Thanks for reply. Did not know it's actually implementation of something.

So the way to catch the warning is indeed to overwrite the error_handler.

-------------

Just to be sure. The warning occurs for me when timeout is reached. Default `timeout` `60` is not enough. When I set value to `120` the request succeeds.

So when 60 is set, the warning is triggered with error message:

> Idle timeout reached for "my endpoint".

Is that expected that it's triggered even for timeouts instead of exception? Or if I ask differently, shouldn't the timeout be handled somewhere else so it can be caught?

**Comment by derrabus** [comment]
Do you interact with the stream wrapper directly or via PHP's stream API? If not, the warning should probably be caught for you. Can you provide a script that reproduces the warning being triggered?

I've tried this piece of code, but it triggers a proper exception.

```php
set_error_handler('dd', E_ALL);
$response = HttpClient::create()
    ->request('GET', 'http://httpbin.org/delay/10', ['timeout' => 5]);
dump($response->getContent());
```

**Comment by simPod** [comment]
<details>

<summary>The client is created via DI</summary>

```xml
<service id="MyNs\HttpClientFactory">
    <argument key="$options" type="collection">
        <argument key="verify_host">false</argument>
        <argument key="verify_peer">false</argument>
        <argument key="extra" type="collection">
            <argument key="curl" type="expression">{(constant("CURLOPT_FORBID_REUSE")): true}</argument>
        </argument>
    </argument>
</service>

<service id="api.http_client" class="Symfony\Component\HttpClient\Psr18Client">
    <factory
        service="MyNs\HttpClientFactory"
        method="create"
    />
</service>
```

```php
final class HttpClientFactory
{
    /** @var array<string, mixed> */
    private array $options;

    /** @param array<string, mixed> $options */
    public function __construct(array $options)
    {
        $this->options = $options;
    }

    public function create() : Psr18Client
    {
        return new Psr18Client(new CurlHttpClient($this->options));
    }
}
```
</details>

![image](https://user-images.githubusercontent.com/327717/125652198-01b326bf-4e69-450e-a468-67496eeabecd.png)

And this is the user land code:

```php
try {
    $response = $this->client->sendRequest($request);
} catch (NetworkException $exception) {
    throw MyExc::fromThrowable($exception);
}

if ($response->getStatusCode() !== Response::HTTP_OK) {
    throw MyExc::unexpectedHttpStatusCode($response->getStatusCode());
}

json_decode((string) $response->getBody(), true); // This triggers the error in stream wrapper on timeout
```

**Comment by derrabus** [comment]
Where is `$message` coming from?

**Comment by simPod** [comment]
Fixed the snippet. It was $response

**Comment by derrabus** [comment]
Another try, this time with the PSR-18 client:

```php
namespace App;

use Nyholm\Psr7\Factory\Psr17Factory;
use Symfony\Component\HttpClient\HttpClient;
use Symfony\Component\HttpClient\Psr18Client;

require __DIR__.'/vendor/autoload.php';

set_error_handler('dd', E_ALL);

$factory = new Psr17Factory();
$client = new Psr18Client(HttpClient::create(['timeout' => 5]), $factory);
$response = $client->sendRequest(
    $factory->createRequest('GET', 'http://httpbin.org/delay/10')
);

dump((string) $response->getBody());
```

I still get an exception instead of a proper warning.

Without a proper reproducer, there isn't much anybody can do for you, I'm afraid.

**Comment by simPod** [comment]
Understand, will try to create a simple app. What is the exception you're getting?

**Comment by derrabus** [comment]
> Understand, will try to create a simple app. What is the exception you're getting?

```
Symfony\Component\HttpClient\Exception\TransportException: Idle timeout reached for "http://httpbin.org/delay/10".
```

**Comment by simPod** [comment]
Hmm, interesting. With httpbin, I'm encountering the same behaviour as you. Might be related to request body or/and the device at the end which you can imagine as a kind of Tamagotchi. Will investigate flow path/get something reproducible and get back. Thanks for your time!

**Comment by simPod** [comment]
The issue is that my Tamagotchi returns headers instantly and then it waits for the body (it takes a long time for it to send it) while httpbin just timeouts the whole request.

So with Tamagotchi there's FirstChunk created with content `''`, then DataChunk with content `{"jsonrpc": "2.0", "id": 1 ` and then the next chunk is not returned in time so cURL WRITEFUNCTION is not called yet, `$multi->handlesActivity[$j]` is not set and no-op is not performed:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/HttpClient/Response/TransportResponseTrait.php#L177-L178

this is triggered instead:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/HttpClient/Response/TransportResponseTrait.php#L182-L183

I still struggle to somehow reproduce it. Did not figure out how to simulate response streaming yet.

**Comment by simPod** [comment]
Ok, I still don't know how to fix it but this reproduces what's happening to me

https://github.com/simPod/symfony/commit/11d1688bb8ff286903dd8344fd3eee3b8d319bdf

**Comment by nicolas-grekas** [comment]
> testTimeoutDuringBodyIsNotAFatalError

But a timeout during body retrieval have to be an error, isn't it? What behavior would you expect otherwise?

**Comment by simPod** [comment]
@nicolas-grekas exception that can be caught, same as with other timeouts. Why should this timeout behave differently - trigger_error, instead of throwing proper exception?

_I've created the test to reproduce the behaviour, it might need a bit of refactoring and rename._

**Comment by nicolas-grekas** [comment]
Thanks for the reproducer. Actually, if you look at the stack trace, the issue is found in nyholm/psr7 (guzzle/psr7 has the same issue):
nyholm/psr7/src/Stream.php:266 doesn't handle warnings potentially thrown by `stream_get_contents()`
(same for other stream functions btw, eg fread(), etc).
Closing therefor, please report there.

**Comment by simPod** [comment]
Thanks for reply. So for recap in Symfony HttpClient:

- TimeoutException should be thrown when request times out and headers are not received
- TimeoutException should NOT be thrown when request times out and some headers were already received
- error with level Warning should be triggered instead of exception when request times out and some header was already received

you see, the only thing that worried be here is that for some timeouts there's exception thrown and for others there's error triggered that can only be caught by overriding error_handler.

Also, I ack that the stream should throw error as @derrabus already stated https://github.com/symfony/symfony/issues/42100#issuecomment-879932476 but I wonder whether this should be already handled outside the stream.

So in conclusion if I get you right, libs that call `stream_get_contents()` should wrap the call in something like

```php
        try {
            set_error_handler(
                static function (int $_, string $errorMessage) : bool {
                    if (error_reporting() === 0) {
                        return false;
                    }

                    throw new \RuntimeException($errorMessage);
                },
                E_USER_WARNING,
            );

            \stream_get_contents($stream)
        } finally {
            restore_error_handler();
        }
```

**Comment by nicolas-grekas** [comment]
> libs that call stream_get_contents() should wrap the call in something like

almost yes (I would not call error_reporting() because here we always want to report the error, and I would pass any deprecations to the previous error handler, if any.)

Or they can skip that and add `@` in front of the call they do (but then I'd hope they'd use `error_get_last()` to provide useful error messages.

**Comment by nicolas-grekas** [comment]
See https://github.com/Nyholm/psr7/pull/185

**Comment by simPod** [comment]
Just returned from a lunch to start working on it to find out you were faster. Thank you!
